### PR TITLE
[sharded_tensor] enable gloo/mpi backend to be used in sharded tensor tests

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66351
* #64481
* __->__ #65855

This adjusted our test base to support non-nccl backend like gloo/mpi, so that we could test sharding on CPU with gloo/mpi backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/testing/_internal/distributed/_sharded_tensor/__init__.py [hunk 1]
```diff
@@ -15,13 +15,30 @@ class ShardedTensorTestBase(MultiProcessTestCase):
...
 
-    def init_pg(self):
-        dist.init_process_group(
-            backend="nccl",
-            world_size=self.world_size,
-            rank=self.rank,
-            init_method=f"file://{self.file_name}",
-        )
+    def init_pg(self, backend="nccl"):
+        if backend == "nccl":
```
Comment (bowangbj) [inline torch/testing/_internal/distributed/_sharded_tensor/__init__.py:19]: probably just 

if backend not in ["nccl", "gloo", "mpi"]  --> raise error 

and then 

dist.init_process_group(
                backend=backend,
                world_size=self.world_size,
                rank=self.rank,
                init_method=f"file://{self.file_name}",
            )

**Review by bowangbj — COMMENTED** [review]
Thanks Wanchao for the fix, just some minior stying issue

**Review by pritamdamania87 — COMMENTED** [review]
This PR just adds the ability to run the tests on different backends but doesn't actual enable running tests on different backends yet right? Are you planning to have a separate PR to do that or are you planning to update this PR to also run all tests with all backends?

**Comment by wanchaol** [comment]
> This PR just adds the ability to run the tests on different backends but doesn't actual enable running tests on different backends yet right? Are you planning to have a separate PR to do that or are you planning to update this PR to also run all tests with all backends?

good point, I think this just enable the ability as you said to unblock tests in init_from_local_shards improvement, I will create a follow up PR to make sure we run all tests with different backends, will also rename this PR

**Review by pritamdamania87 — APPROVED** [review]

**Review by bowangbj — APPROVED** [review]
LGTM, thx Wanchao

**Pull request closed without merging**